build.rs

```diff
@@ -23,85 +73,72 @@ fn probe_r_paths() -> io::Result<InstallationPaths> {
-
-            match lines.next() {
-                Some(line) => line.to_string(),
-                _ => return Err(Error::new(ErrorKind::Other, "Cannot find R home.")),
+            let rout = byte_array_to_os_string(&rout.stdout);
+            if !rout.is_empty() {
+                PathBuf::from(rout)
+            } else {
+                return Err(Error::new(ErrorKind::Other, "Cannot find R home."));
             }
         }
     };
 
     // Now the library location. On Windows, it depends on the target architecture
-    let pkg_target_arch = env::var("CARGO_CFG_TARGET_ARCH").unwrap();
+    let pkg_target_arch = env::var_os("CARGO_CFG_TARGET_ARCH").unwrap();
     let library = if cfg!(target_os = "windows") {
         if pkg_target_arch == "x86_64" {
             Path::new(&r_home)
                 .join("bin")
                 .join("x64")
-                .to_str()
-                .unwrap()
-                .to_string()
         } else if pkg_target_arch == "x86" {
             Path::new(&r_home)
                 .join("bin")
                 .join("i386")
-                .to_str()
-                .unwrap()
-                .to_string()
         } else {
             panic!("Unknown architecture")
         }
     } else {
-        Path::new(&r_home).join("lib").to_str().unwrap().to_string()
+        Path::new(&r_home).join("lib")
     };
 
     // Finally the include location. It may or may not be located under R home
-    let include = match env::var("R_INCLUDE_DIR") {
+    let include = match env::var_os("R_INCLUDE_DIR") {
         // If the environment variable R_INCLUDE_DIR is set we use it
-        Ok(s) => s,
+        Some(s) => PathBuf::from(s),
 
         // Otherwise, we try to execute `R` to find the include dir. Here,
         // we're using the R home we found earlier, to make sure we're consistent.
         _ => {
             let r_binary = if cfg!(target_os = "windows") {
                 Path::new(&library)
                     .join("R.exe")
-                    .to_str()
-                    .unwrap()
-                    .to_string()
             } else {
                 Path::new(&r_home)
                     .join("bin")
                     .join("R")
-                    .to_str()
-                    .unwrap()
-                    .to_string()
             };
 
             let out = Command::new(&r_binary)
                 .args(&[
                     "-s",
                     "-e",
-                    r#"cat(normalizePath(R.home('include')), sep = '\n')"#
+                    r#"cat(normalizePath(R.home('include')))"#
                 ])
                 .output()?;
 
             // if there are any errors we print them out, helps with debugging
-            for errln in String::from_utf8_lossy(&out.stderr).lines() {
-                println!("> {}", errln);
+            if !out.stderr.is_empty() {
+                println!("> {}", String::from_utf8_lossy(&out.stderr));
```

Comment:
Can we use `byte_array_to_os_string()` here as well?

Reply:
I don't think so. (Or at least I don't know how.) `OsString` doesn't implement the `Display` trait, so I can't print it. In the rest of the code I got around this issue by using `PathBuf`, which does implement `Display`. But we can't convert an arbitrary message into a `PathBuf`. I really don't understand why `OsString` doesn't implement `Display`.

Reply:
I meant something like `byte_array_to_os_string(&out.stderr).as_os_str().to_string_lossy()`. `String::from_utf8_lossy(&out.stderr)` is trying to interpret the non-UTF-8 byte array as UTF-8, which seems wrong, if I understand correctly.

Reply:
Ah, I see. Yes, it's possible that this is somewhat more reliable. I'll make the change.